File: Backend/src/main/java/com/luckyseven/backend/domain/budget/entity/Budget.java
package com.luckyseven.backend.domain.budget.entity;

import static com.luckyseven.backend.sharedkernel.exception.ExceptionCode.INSUFFICIENT_BALANCE;

import com.luckyseven.backend.domain.team.entity.Team;
import com.luckyseven.backend.sharedkernel.entity.BaseEntity;
import com.luckyseven.backend.sharedkernel.exception.CustomLogicException;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToOne;
import java.math.BigDecimal;
import java.math.RoundingMode;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Budget extends BaseEntity {

  @Id
  @Column(name = "budget_id")
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @OneToOne(mappedBy = "budget")
  private Team team;

  @Column(nullable = false)
  private BigDecimal totalAmount;

  @Setter
  @Column(nullable = false)
  private Long setBy;

  @Column(nullable = false)
  private BigDecimal balance;

  private BigDecimal foreignBalance;

  @Setter
  @Enumerated(EnumType.STRING)
  @Column(nullable = false, length = 3)
  private CurrencyCode foreignCurrency;

  private BigDecimal avgExchangeRate;

  @Builder
  public Budget(Team team, BigDecimal totalAmount, Long setBy,
      BigDecimal balance, BigDecimal foreignBalance, CurrencyCode foreignCurrency,
      BigDecimal avgExchangeRate) {
    this.team = team;
    this.totalAmount = totalAmount;
    this.setBy = setBy;
    this.balance = balance;
    this.foreignBalance = foreignBalance;
    this.foreignCurrency = foreignCurrency;
    this.avgExchangeRate = avgExchangeRate;
  }

  public void setTotalAmount(BigDecimal totalAmount) {
    this.balance = this.balance.add(totalAmount.subtract(this.totalAmount));
    this.totalAmount = totalAmount;
  }

  public void setExchangeInfo(boolean isExchanged, BigDecimal amount, BigDecimal exchangeRate) {
    if (!isExchanged) {
      this.avgExchangeRate = null;
      this.foreignBalance = null;
      return;
    }

    updateForeignBalance(amount, exchangeRate);
  }

  public void updateExchangeInfo(boolean isExchanged, BigDecimal amount, BigDecimal exchangeRate) {
    if (!isExchanged) {
      return;
    }

    updateAvgExchangeRate(amount, exchangeRate);
    updateForeignBalance(amount, exchangeRate);

  }

  // 예산 추가 후 외화잔고 및 평균환율 수정
  private void updateAvgExchangeRate(BigDecimal amount, BigDecimal exchangeRate) {
    if (this.avgExchangeRate == null || this.avgExchangeRate.compareTo(BigDecimal.ZERO) == 0) {
      avgExchangeRate = exchangeRate;
      return;
    }
    BigDecimal foreignAmount = amount.divide(exchangeRate, 10,
        RoundingMode.HALF_UP); // 외화 환산, 충분한 정밀도 확보
    BigDecimal totalCost = this.foreignBalance.multiply(this.avgExchangeRate).add(amount);
    BigDecimal totalForeign = this.foreignBalance.add(foreignAmount);
    this.avgExchangeRate = totalCost.divide(totalForeign, 2, RoundingMode.HALF_UP);

  }

  private void updateForeignBalance(BigDecimal amount, BigDecimal exchangeRate) {
    BigDecimal additionalBudget = amount.divide(exchangeRate, 2, RoundingMode.HALF_UP);
    if (this.foreignBalance == null) {
      foreignBalance = BigDecimal.ZERO;
    }
    this.foreignBalance = this.foreignBalance.add(additionalBudget);
    this.avgExchangeRate = exchangeRate;
  }

  public void setForeignBalance() {
    if (avgExchangeRate != null) {
      this.foreignBalance = totalAmount.divide(avgExchangeRate, 2, RoundingMode.HALF_UP);
    }
  }

  public void setForeignBalance(BigDecimal amount) {
    this.foreignBalance = amount;
  }

  public Budget setTeam(Team team) {
    // 기존 연결 해제
    if (this.team != null) {
      this.team.setBudget(null);
    }

    this.team = team;

    // 새로운 연결 설정 (Team이 null이 아닌 경우)
    if (team != null && team.getBudget() != this) {
      team.setBudget(this);
    }

    return this;
  }

  public void debitKrw(BigDecimal krwAmount) {
    validateSufficientBalance(krwAmount, balance);
    balance = balance.subtract(krwAmount);
  }

  public void debitForeign(BigDecimal foreignAmount) {
    BigDecimal krwAmount = foreignAmount.multiply(avgExchangeRate);
    validateSufficientBalance(krwAmount, balance);
    validateSufficientBalance(foreignAmount, foreignBalance);
    balance = balance.subtract(krwAmount);
    foreignBalance = foreignBalance.subtract(foreignAmount);
  }

  public void creditKrw(BigDecimal krwAmount) {
    balance = balance.add(krwAmount);
  }

  public void creditForeign(BigDecimal foreignAmount) {
    BigDecimal krwAmount = foreignAmount.multiply(avgExchangeRate);
    this.balance = this.balance.add(krwAmount);

    if (this.foreignBalance == null) {
      this.foreignBalance = BigDecimal.ZERO;
    }

    this.foreignBalance = this.foreignBalance.add(foreignAmount);
  }


  private void validateSufficientBalance(BigDecimal amount, BigDecimal current) {
    if (current.compareTo(amount) < 0) {
      throw new CustomLogicException(INSUFFICIENT_BALANCE);
    }
  }

}

File: Backend/src/main/kotlin/com/luckyseven/backend/domain/budget/entity/Budget.kt
package com.luckyseven.backend.domain.budget.entity

import com.luckyseven.backend.domain.team.entity.Team
import com.luckyseven.backend.sharedkernel.entity.BaseEntity
import com.luckyseven.backend.sharedkernel.exception.CustomLogicException
import com.luckyseven.backend.sharedkernel.exception.ExceptionCode
import jakarta.persistence.*
import java.math.BigDecimal
import java.math.RoundingMode

@Entity
@Table(
    indexes = [
        Index(name = "idx_budget_team", columnList = "team_id")
    ]
)
class Budget(

    @Id
    @Column(name = "budget_id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @OneToOne(mappedBy = "budget", fetch = FetchType.LAZY)
    var team: Team? = null,

    totalAmount: BigDecimal,

    @Column(nullable = false)
    var setBy: Long,

    @Column(nullable = false)
    var balance: BigDecimal,

    @Column
    var foreignBalance: BigDecimal? = null,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 3)
    val foreignCurrency: CurrencyCode,

    @Column
    var avgExchangeRate: BigDecimal? = null

) : BaseEntity() {

    companion object {
        private const val SCALE = 2
        private val ROUNDING: RoundingMode = RoundingMode.HALF_UP
        private val ZERO = BigDecimal.ZERO
    }

    @Column(nullable = false)
    var totalAmount: BigDecimal = totalAmount
        set(value) {
            balance = balance.add(value.subtract(field))
            field = value
        }

    fun setExchangeInfo(isExchanged: Boolean, amount: BigDecimal, exchangeRate: BigDecimal?) {
        if (!isExchanged) {
            this.avgExchangeRate = null
            this.foreignBalance = null
            return
        }

        updateForeignBalance(amount, exchangeRate!!)
    }

    fun updateExchangeInfo(isExchanged: Boolean, amount: BigDecimal, exchangeRate: BigDecimal?) {
        if (!isExchanged) {
            return
        }

        // isExchanged가 true면 exchangeRate를 반드시 포함
        updateAvgExchangeRate(amount, exchangeRate!!)
        updateForeignBalance(amount, exchangeRate!!)
    }

    // 예산 추가 후 외화잔고 및 평균환율 수정
    private fun updateAvgExchangeRate(amount: BigDecimal, exchangeRate: BigDecimal) {
        if (this.avgExchangeRate == null || this.avgExchangeRate!!.compareTo(ZERO) == 0) {
            avgExchangeRate = exchangeRate
            return
        }
        val foreignAmount = amount.divide(
            exchangeRate, 10,
            ROUNDING
        ) // 외화 환산, 충분한 정밀도 확보
        val totalCost = this.foreignBalance!!.multiply(this.avgExchangeRate).add(amount)
        val totalForeign = this.foreignBalance!!.add(foreignAmount)
        this.avgExchangeRate = totalCost.divide(totalForeign, SCALE, ROUNDING)
    }

    private fun updateForeignBalance(amount: BigDecimal, exchangeRate: BigDecimal) {
        val additionalBudget = amount.divide(exchangeRate, SCALE, ROUNDING)
        if (this.foreignBalance == null) {
            foreignBalance = ZERO
        }
        this.foreignBalance = this.foreignBalance!!.add(additionalBudget)
        this.avgExchangeRate = exchangeRate
    }

    fun setForeignBalance() {
        if (avgExchangeRate != null) {
            this.foreignBalance = balance.divide(avgExchangeRate, SCALE, ROUNDING)
        }
    }

    fun setTeam(team: Team?): Budget {
        // 기존 연결 해제
        this.team?.budget = null

        this.team = team

        // 새로운 연결 설정 (Team이 null이 아닌 경우)
        if (team != null && team.budget !== this) {
            team.budget = this
        }

        return this
    }

    fun debitKrw(krwAmount: BigDecimal) {
        validateSufficientBalance(krwAmount, balance)
        balance = balance.subtract(krwAmount)
    }

    fun debitForeign(foreignAmount: BigDecimal) {
        val krwAmount = foreignAmount.multiply(avgExchangeRate)
        validateSufficientBalance(krwAmount, balance)
        validateSufficientBalance(foreignAmount, foreignBalance!!)
        balance = balance.subtract(krwAmount)
        foreignBalance = foreignBalance!!.subtract(foreignAmount)
    }

    fun creditKrw(krwAmount: BigDecimal) {
        balance = balance.add(krwAmount)
    }

    fun creditForeign(foreignAmount: BigDecimal) {
        val krwAmount = foreignAmount.multiply(avgExchangeRate)
        this.balance = this.balance.add(krwAmount)

        if (this.foreignBalance == null) {
            this.foreignBalance = ZERO
        }

        this.foreignBalance = this.foreignBalance!!.add(foreignAmount)
    }


    private fun validateSufficientBalance(amount: BigDecimal?, current: BigDecimal) {
        if (current < amount) {
            throw CustomLogicException(ExceptionCode.INSUFFICIENT_BALANCE)
        }
    }
}
